File: sorting/selection_sort.py
import math


# selection_sort
def selection_sort(arr):

    for i in range(len(arr)-1):

        current_min = arr[i]
        min_idx = i
        for j in range(i+1, len(arr)):

            # find minimum
            if arr[j] < current_min:
                current_min = arr[j]
                min_idx = j

        # swap minimum with i'th element
        arr[i], arr[min_idx] = arr[min_idx], arr[i]


def main():

    arr = [5,6,8,1,2,4,5,10]
    selection_sort(arr)

    print(f'{arr}')


if __name__ == "__main__":
    main()

File: sorting/selectionsort.go
package main

import "log"

// returns the minimum in given sequence
func minimumIndex(seq []int) int {

	min := seq[0]
	minIdx := 0
	for i, v := range seq {
		if v < min {
			min = v
			minIdx = i
		}
	}

	return minIdx
}

// selectionSort - sorts given sequence in place
func selectionSort(seq []int) {

	for i:=0; i<len(seq)-1; i++ {

		// find minimum on each iteration
		minIdx := minimumIndex(seq[i:])+i

		// place minimum to the correct slot in the sequence
		seq[minIdx], seq[i] = seq[i], seq[minIdx]
	}
}


func main() {
	seq := []int{5,6,1,2,6,2}
	selectionSort(seq)

	log.Printf("after %v", seq)
}
